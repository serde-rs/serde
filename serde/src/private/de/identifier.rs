use crate::lib::*;

use crate::de::{DeserializeSeed, Deserializer, Error, Unexpected, Visitor};
use crate::private::from_utf8_lossy;
#[cfg(any(feature = "std", feature = "alloc"))]
use crate::serde_core_private::Content;

// FIXME: this is a copy of serde_core::format::Buf because it is private
struct Buf<'a> {
    bytes: &'a mut [u8],
    offset: usize,
}

impl<'a> Buf<'a> {
    pub fn new(bytes: &'a mut [u8]) -> Self {
        Buf { bytes, offset: 0 }
    }

    pub fn as_str(&self) -> &str {
        let slice = &self.bytes[..self.offset];
        unsafe { str::from_utf8_unchecked(slice) }
    }
}

impl<'a> fmt::Write for Buf<'a> {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        if self.offset + s.len() > self.bytes.len() {
            Err(fmt::Error)
        } else {
            self.bytes[self.offset..self.offset + s.len()].copy_from_slice(s.as_bytes());
            self.offset += s.len();
            Ok(())
        }
    }
}

////////////////////////////////////////////////////////////////////////////////

/// Represents field of a struct
#[derive(Debug)]
pub enum Field {
    /// Represents non-skipped field with specified index assigned to the field
    /// after skipping all non-deserializable fields. For example, for a struct
    ///
    /// ```ignore
    /// #[derive(Deserialize)]
    /// struct StructWithSkippedFields {
    ///     #[serde(skip_deserializing)]
    ///     skipped: (),
    ///
    ///     field: (),
    /// }
    /// ```
    /// field `field` will have an index `0`, because it is the first non-skipped
    /// field of the struct (and counting starts with zero).
    Field(usize),
    /// Field that is not present in struct
    Unknown,
}

/// Creates a field deserialization seed that wrapped array with all possible
/// field names and their aliases.
///
/// # Example
///
/// ```ignore
/// let seed = FieldSeed::new(&[
///     // First field with two aliases
///     &["a", "alias 1", "alias 2"],
///     // Second field with one alias
///     &["b", "alias 3"],
///     // Third field without aliases
///     &["c"],
/// ]);
/// ```
#[derive(Debug)]
pub struct FieldSeed<'a> {
    /// List of all field aliases in order of their presence in the struct
    aliases: &'a [&'a [&'a str]],
}

impl<'a> FieldSeed<'a> {
    #[allow(missing_docs)]
    pub const fn new(aliases: &'a [&'a [&'a str]]) -> Self {
        Self { aliases }
    }

    fn matches(&self, value: &[u8]) -> Option<usize> {
        self.aliases
            .iter()
            .position(|aliases| aliases.iter().any(|a| a.as_bytes() == value))
    }

    fn index(&self, value: u64) -> Field {
        if value < self.aliases.len() as u64 {
            Field::Field(value as usize)
        } else {
            Field::Unknown
        }
    }
}

impl<'de, 'a, 'b> Visitor<'de> for &'b mut FieldSeed<'a> {
    type Value = Field;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("field identifier")
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(self.index(value))
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        self.visit_bytes(value.as_bytes())
    }

    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value) {
            Some(i) => Ok(Field::Field(i)),
            None => Ok(Field::Unknown),
        }
    }
}

impl<'de, 'a, 'b> DeserializeSeed<'de> for &'b mut FieldSeed<'a> {
    type Value = Field;

    #[inline]
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_identifier(self)
    }
}

////////////////////////////////////////////////////////////////////////////////

/// Represents field of a struct that denies unknown fields
#[derive(Debug)]
pub enum FieldStrong {
    /// Represents non-skipped field with specified index assigned to the field
    /// after skipping all non-deserializable fields. For example, for a struct
    ///
    /// ```ignore
    /// #[derive(Deserialize)]
    /// struct StructWithSkippedFields {
    ///     #[serde(skip_deserializing)]
    ///     skipped: (),
    ///
    ///     field: (),
    /// }
    /// ```
    /// field `field` will have an index `0`, because it is the first non-skipped
    /// field of the struct (and counting starts with zero).
    Field(usize),
}

/// Creates a field deserialization seed that wrapped array with all possible
/// field names and their aliases.
///
/// # Example
///
/// ```ignore
/// let seed = FieldStrongSeed::new(
///     &[
///         // First field with two aliases
///         &["a", "alias 1", "alias 2"],
///         // Second field with one alias
///         &["b", "alias 3"],
///         // Third field without aliases
///         &["c"],
///     ],
///     &[
///         // First field with two aliases
///         "a", "alias 1", "alias 2",
///         // Second field with one alias
///         "b", "alias 3",
///         // Third field without aliases
///         "c",
///     ],
/// );
/// ```
#[derive(Debug)]
pub struct FieldStrongSeed<'a> {
    seed: FieldSeed<'a>,
    fields: &'static [&'static str],
}

impl<'a> FieldStrongSeed<'a> {
    #[allow(missing_docs)]
    pub const fn new(aliases: &'a [&'a [&'a str]], fields: &'static [&'static str]) -> Self {
        Self {
            seed: FieldSeed::new(aliases),
            fields,
        }
    }

    fn matches(&self, value: &[u8]) -> Option<FieldStrong> {
        match self.seed.matches(value) {
            Some(i) => Some(FieldStrong::Field(i)),
            None => None,
        }
    }
}

impl<'de, 'a, 'b> Visitor<'de> for &'b mut FieldStrongSeed<'a> {
    type Value = FieldStrong;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("field identifier")
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.seed.index(value) {
            Field::Field(i) => Ok(FieldStrong::Field(i)),
            Field::Unknown => {
                // length of string "field index 0 <= i < " and u64::MAX.to_string()
                let mut buf = [0u8; 21 + 20];
                let mut writer = Buf::new(&mut buf);
                fmt::Write::write_fmt(&mut writer, format_args!("field index 0 <= i < {}", value))
                    .unwrap();
                Err(Error::invalid_value(
                    Unexpected::Unsigned(value),
                    &writer.as_str(),
                ))
            }
        }
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value.as_bytes()) {
            Some(field) => Ok(field),
            None => Err(Error::unknown_field(value, self.fields)),
        }
    }

    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value) {
            Some(field) => Ok(field),
            None => Err(Error::unknown_field(&from_utf8_lossy(value), self.fields)),
        }
    }
}

impl<'de, 'a, 'b> DeserializeSeed<'de> for &'b mut FieldStrongSeed<'a> {
    type Value = FieldStrong;

    #[inline]
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_identifier(self)
    }
}

////////////////////////////////////////////////////////////////////////////////

/// Represents field of a struct with flatten fields
#[cfg(any(feature = "std", feature = "alloc"))]
pub enum FieldFlatten<'de> {
    /// Represents non-skipped field with specified index assigned to the field
    /// after skipping all non-deserializable fields. For example, for a struct
    ///
    /// ```ignore
    /// #[derive(Deserialize)]
    /// struct StructWithSkippedFields {
    ///     #[serde(skip_deserializing)]
    ///     skipped: (),
    ///
    ///     field: (),
    /// }
    /// ```
    /// field `field` will have an index `0`, because it is the first non-skipped
    /// field of the struct (and counting starts with zero).
    Field(usize),
    /// Field that is not present in the struct and will be forwarded to the
    /// flatten fields
    Other(Content<'de>),
}

/// Creates a field deserialization seed that wrapped array with all possible
/// field names and their aliases.
///
/// # Example
///
/// ```ignore
/// let seed = FieldFlattenSeed::new(&[
///     // First field with two aliases
///     &["a", "alias 1", "alias 2"],
///     // Second field with one alias
///     &["b", "alias 3"],
///     // Third field without aliases
///     &["c"],
/// ]);
/// ```
#[cfg(any(feature = "std", feature = "alloc"))]
#[derive(Debug)]
pub struct FieldFlattenSeed<'a> {
    seed: FieldSeed<'a>,
}

#[cfg(any(feature = "std", feature = "alloc"))]
impl<'a> FieldFlattenSeed<'a> {
    #[allow(missing_docs)]
    pub const fn new(aliases: &'a [&'a [&'a str]]) -> Self {
        Self {
            seed: FieldSeed::new(aliases),
        }
    }

    fn matches(&self, value: &[u8]) -> Option<FieldFlatten<'static>> {
        match self.seed.matches(value) {
            Some(i) => Some(FieldFlatten::Field(i)),
            None => None,
        }
    }
}

#[cfg(any(feature = "std", feature = "alloc"))]
impl<'de, 'a, 'b> Visitor<'de> for &'b mut FieldFlattenSeed<'a> {
    type Value = FieldFlatten<'de>;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("field identifier")
    }

    fn visit_bool<E>(self, value: bool) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::Bool(value)))
    }

    fn visit_i8<E>(self, value: i8) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::I8(value)))
    }

    fn visit_i16<E>(self, value: i16) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::I16(value)))
    }

    fn visit_i32<E>(self, value: i32) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::I32(value)))
    }

    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::I64(value)))
    }

    fn visit_u8<E>(self, value: u8) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::U8(value)))
    }

    fn visit_u16<E>(self, value: u16) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::U16(value)))
    }

    fn visit_u32<E>(self, value: u32) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::U32(value)))
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::U64(value)))
    }

    fn visit_f32<E>(self, value: f32) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::F32(value)))
    }

    fn visit_f64<E>(self, value: f64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::F64(value)))
    }

    fn visit_char<E>(self, value: char) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::Char(value)))
    }

    fn visit_unit<E>(self) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(FieldFlatten::Other(Content::Unit))
    }

    fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value.as_bytes()) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::String(value))),
        }
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value.as_bytes()) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::String(value.to_string()))),
        }
    }

    fn visit_borrowed_str<E>(self, value: &'de str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value.as_bytes()) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::Str(value))),
        }
    }

    fn visit_byte_buf<E>(self, value: Vec<u8>) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(&value) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::ByteBuf(value))),
        }
    }

    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::ByteBuf(value.to_vec()))),
        }
    }

    fn visit_borrowed_bytes<E>(self, value: &'de [u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value) {
            Some(field) => Ok(field),
            None => Ok(FieldFlatten::Other(Content::Bytes(value))),
        }
    }
}

#[cfg(any(feature = "std", feature = "alloc"))]
impl<'de, 'a, 'b> DeserializeSeed<'de> for &'b mut FieldFlattenSeed<'a> {
    type Value = FieldFlatten<'de>;

    #[inline]
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_identifier(self)
    }
}

////////////////////////////////////////////////////////////////////////////////

/// Creates a field deserialization seed that wrapped array with all possible
/// field names and their aliases.
///
/// # Example
///
/// ```ignore
/// let seed = VariantSeed::new(
///     &[
///         // First field with two aliases
///         &["a", "alias 1", "alias 2"],
///         // Second field with one alias
///         &["b", "alias 3"],
///         // Third field without aliases
///         &["c"],
///     ],
///     &[
///         // First field with two aliases
///         "a", "alias 1", "alias 2",
///         // Second field with one alias
///         "b", "alias 3",
///         // Third field without aliases
///         "c",
///     ],
/// );
/// ```
#[derive(Debug)]
pub struct VariantSeed<'a> {
    aliases: &'a [&'a [&'a str]],
    variants: &'static [&'static str],
}

impl<'a> VariantSeed<'a> {
    #[allow(missing_docs)]
    pub const fn new(aliases: &'a [&'a [&'a str]], variants: &'static [&'static str]) -> Self {
        Self { aliases, variants }
    }

    fn matches(&self, value: &[u8]) -> Option<usize> {
        self.aliases
            .iter()
            .position(|variant| variant.iter().any(|v| v.as_bytes() == value))
    }
}

impl<'de, 'a> Visitor<'de> for VariantSeed<'a> {
    type Value = usize;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("variant identifier")
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        if value < self.aliases.len() as u64 {
            Ok(value as usize)
        } else {
            // length of string "variant index 0 <= i < " and u64::MAX.to_string()
            let mut buf = [0u8; 23 + 20];
            let mut writer = Buf::new(&mut buf);
            fmt::Write::write_fmt(
                &mut writer,
                format_args!("variant index 0 <= i < {}", value),
            )
            .unwrap();
            Err(Error::invalid_value(
                Unexpected::Unsigned(value),
                &writer.as_str(),
            ))
        }
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value.as_bytes()) {
            Some(index) => Ok(index),
            None => Err(Error::unknown_variant(value, self.variants)),
        }
    }

    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        match self.matches(value) {
            Some(index) => Ok(index),
            None => Err(Error::unknown_variant(
                &from_utf8_lossy(value),
                self.variants,
            )),
        }
    }
}

impl<'de, 'a> DeserializeSeed<'de> for VariantSeed<'a> {
    type Value = usize;

    #[inline]
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_identifier(self)
    }
}

////////////////////////////////////////////////////////////////////////////////

/// Creates a field deserialization seed that wrapped array with all possible
/// field names and their aliases.
///
/// # Example
///
/// ```ignore
/// let seed = VariantOtherSeed::new(
///     &[
///         // First field with two aliases
///         &["a", "alias 1", "alias 2"],
///         // Second field with one alias
///         &["b", "alias 3"],
///         // Third field without aliases
///         &["c"],
///     ],
///     2,
/// );
/// ```
#[derive(Debug)]
pub struct VariantOtherSeed<'a> {
    aliases: &'a [&'a [&'a str]],
    other: usize,
}

impl<'a> VariantOtherSeed<'a> {
    #[allow(missing_docs)]
    pub const fn new(aliases: &'a [&'a [&'a str]], other: usize) -> Self {
        Self { aliases, other }
    }

    fn matches(&self, value: &[u8]) -> usize {
        self.aliases
            .iter()
            .position(|variant| variant.iter().any(|v| v.as_bytes() == value))
            .unwrap_or(self.other)
    }
}

impl<'de, 'a> Visitor<'de> for VariantOtherSeed<'a> {
    type Value = usize;

    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("variant identifier")
    }

    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
    where
        E: Error,
    {
        if value < self.aliases.len() as u64 {
            Ok(value as usize)
        } else {
            Ok(self.other)
        }
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(self.matches(value.as_bytes()))
    }

    fn visit_bytes<E>(self, value: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        Ok(self.matches(value))
    }
}

impl<'de, 'a> DeserializeSeed<'de> for VariantOtherSeed<'a> {
    type Value = usize;

    #[inline]
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_identifier(self)
    }
}
