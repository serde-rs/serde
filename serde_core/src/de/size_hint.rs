//! Provides helpers for creating size hints for container deserialization.
#[cfg(any(feature = "std", feature = "alloc"))]
use crate::lib::*;

/// Extracts the exact size of an iterator if it has a known upper bound and it matches the lower bound.
pub fn from_bounds<I>(iter: &I) -> Option<usize>
where
    I: Iterator,
{
    helper(iter.size_hint())
}

/// Returns conservative size estimate for a container, clamping the result to a maximum size.
#[cfg(any(feature = "std", feature = "alloc"))]
pub fn cautious<Element>(hint: Option<usize>) -> usize {
    const MAX_PREALLOC_BYTES: usize = 1024 * 1024;

    if mem::size_of::<Element>() == 0 {
        0
    } else {
        cmp::min(
            hint.unwrap_or(0),
            MAX_PREALLOC_BYTES / mem::size_of::<Element>(),
        )
    }
}

fn helper(bounds: (usize, Option<usize>)) -> Option<usize> {
    match bounds {
        (lower, Some(upper)) if lower == upper => Some(upper),
        _ => None,
    }
}
