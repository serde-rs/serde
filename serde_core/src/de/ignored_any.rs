use crate::lib::*;

use crate::de::{
    Deserialize, Deserializer, EnumAccess, Error, MapAccess, SeqAccess, VariantAccess, Visitor,
};

/// An efficient way of discarding data from a deserializer.
///
/// Think of this like `serde_json::Value` in that it can be deserialized from
/// any type, except that it does not store any information about the data that
/// gets deserialized.
///
/// ```edition2021
/// use serde::de::{
///     self, Deserialize, DeserializeSeed, Deserializer, IgnoredAny, SeqAccess, Visitor,
/// };
/// use std::fmt;
/// use std::marker::PhantomData;
///
/// /// A seed that can be used to deserialize only the `n`th element of a sequence
/// /// while efficiently discarding elements of any type before or after index `n`.
/// ///
/// /// For example to deserialize only the element at index 3:
/// ///
/// /// ```
/// /// NthElement::new(3).deserialize(deserializer)
/// /// ```
/// pub struct NthElement<T> {
///     n: usize,
///     marker: PhantomData<T>,
/// }
///
/// impl<T> NthElement<T> {
///     pub fn new(n: usize) -> Self {
///         NthElement {
///             n: n,
///             marker: PhantomData,
///         }
///     }
/// }
///
/// impl<'de, T> Visitor<'de> for NthElement<T>
/// where
///     T: Deserialize<'de>,
/// {
///     type Value = T;
///
///     fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
///         write!(
///             formatter,
///             "a sequence in which we care about element {}",
///             self.n
///         )
///     }
///
///     fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
///     where
///         A: SeqAccess<'de>,
///     {
///         // Skip over the first `n` elements.
///         for i in 0..self.n {
///             // It is an error if the sequence ends before we get to element `n`.
///             if seq.next_element::<IgnoredAny>()?.is_none() {
///                 return Err(de::Error::invalid_length(i, &self));
///             }
///         }
///
///         // Deserialize the one we care about.
///         let nth = match seq.next_element()? {
///             Some(nth) => nth,
///             None => {
///                 return Err(de::Error::invalid_length(self.n, &self));
///             }
///         };
///
///         // Skip over any remaining elements in the sequence after `n`.
///         while let Some(IgnoredAny) = seq.next_element()? {
///             // ignore
///         }
///
///         Ok(nth)
///     }
/// }
///
/// impl<'de, T> DeserializeSeed<'de> for NthElement<T>
/// where
///     T: Deserialize<'de>,
/// {
///     type Value = T;
///
///     fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
///     where
///         D: Deserializer<'de>,
///     {
///         deserializer.deserialize_seq(self)
///     }
/// }
///
/// # fn example<'de, D>(deserializer: D) -> Result<(), D::Error>
/// # where
/// #     D: Deserializer<'de>,
/// # {
/// // Deserialize only the sequence element at index 3 from this deserializer.
/// // The element at index 3 is required to be a string. Elements before and
/// // after index 3 are allowed to be of any type.
/// let s: String = NthElement::new(3).deserialize(deserializer)?;
/// #     Ok(())
/// # }
/// ```
#[derive(Copy, Clone, Debug, Default, PartialEq)]
pub struct IgnoredAny;

impl<'de> Visitor<'de> for IgnoredAny {
    type Value = IgnoredAny;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("anything at all")
    }

    #[inline]
    fn visit_bool<E>(self, x: bool) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_i64<E>(self, x: i64) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_i128<E>(self, x: i128) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_u64<E>(self, x: u64) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_u128<E>(self, x: u128) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_f64<E>(self, x: f64) -> Result<Self::Value, E> {
        let _ = x;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
    where
        E: Error,
    {
        let _ = s;
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_none<E>(self) -> Result<Self::Value, E> {
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_some<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        IgnoredAny::deserialize(deserializer)
    }

    #[inline]
    fn visit_newtype_struct<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        IgnoredAny::deserialize(deserializer)
    }

    #[inline]
    fn visit_unit<E>(self) -> Result<Self::Value, E> {
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
    where
        A: SeqAccess<'de>,
    {
        while let Some(IgnoredAny) = tri!(seq.next_element()) {
            // Gobble
        }
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
    where
        A: MapAccess<'de>,
    {
        while let Some((IgnoredAny, IgnoredAny)) = tri!(map.next_entry()) {
            // Gobble
        }
        Ok(IgnoredAny)
    }

    #[inline]
    fn visit_bytes<E>(self, bytes: &[u8]) -> Result<Self::Value, E>
    where
        E: Error,
    {
        let _ = bytes;
        Ok(IgnoredAny)
    }

    fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
    where
        A: EnumAccess<'de>,
    {
        tri!(data.variant::<IgnoredAny>()).1.newtype_variant()
    }
}

impl<'de> Deserialize<'de> for IgnoredAny {
    #[inline]
    fn deserialize<D>(deserializer: D) -> Result<IgnoredAny, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_ignored_any(IgnoredAny)
    }
}
