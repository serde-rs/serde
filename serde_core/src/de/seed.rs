use crate::de::{Deserialize, DeserializeSeed, Deserializer};

/// A DeserializeSeed helper for implementing deserialize_in_place Visitors.
///
/// Wraps a mutable reference and calls deserialize_in_place on it.
pub struct InPlaceSeed<'a, T: 'a>(pub &'a mut T);

impl<'a, 'de, T> DeserializeSeed<'de> for InPlaceSeed<'a, T>
where
    T: Deserialize<'de>,
{
    type Value = ();
    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
    where
        D: Deserializer<'de>,
    {
        T::deserialize_in_place(deserializer, self.0)
    }
}
