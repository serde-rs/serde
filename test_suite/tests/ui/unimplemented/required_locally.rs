use serde::de::Deserialize;
use serde::ser::Serialize;

fn to_string<T>(_: &T) -> String
where
    T: Serialize,
{
    unimplemented!()
}

fn from_str<'de, T>(_: &'de str) -> T
where
    T: Deserialize<'de>,
{
    unimplemented!()
}

struct MyStruct;

fn main() {
    to_string(&MyStruct);
    let _: MyStruct = from_str("");
}
