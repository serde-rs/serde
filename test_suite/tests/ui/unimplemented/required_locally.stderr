error[E0277]: the trait bound `MyStruct: serde::Serialize` is not satisfied
  --> tests/ui/unimplemented/required_locally.rs:21:15
   |
21 |     to_string(&MyStruct);
   |     --------- ^^^^^^^^^ unsatisfied trait bound
   |     |
   |     required by a bound introduced by this call
   |
help: the trait `Serialize` is not implemented for `MyStruct`
  --> tests/ui/unimplemented/required_locally.rs:18:1
   |
18 | struct MyStruct;
   | ^^^^^^^^^^^^^^^
   = note: for local types consider adding `#[derive(serde::Serialize)]` to your `MyStruct` type
   = note: for types from other crates check whether the crate offers a `serde` feature flag
   = help: the following other types implement trait `Serialize`:
             &'a T
             &'a mut T
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
           and $N others
note: required by a bound in `to_string`
  --> tests/ui/unimplemented/required_locally.rs:6:8
   |
 4 | fn to_string<T>(_: &T) -> String
   |    --------- required by a bound in this function
 5 | where
 6 |     T: Serialize,
   |        ^^^^^^^^^ required by this bound in `to_string`

error[E0277]: the trait bound `MyStruct: serde::Deserialize<'de>` is not satisfied
  --> tests/ui/unimplemented/required_locally.rs:22:23
   |
22 |     let _: MyStruct = from_str("");
   |                       ^^^^^^^^^^^^ unsatisfied trait bound
   |
help: the trait `Deserialize<'_>` is not implemented for `MyStruct`
  --> tests/ui/unimplemented/required_locally.rs:18:1
   |
18 | struct MyStruct;
   | ^^^^^^^^^^^^^^^
   = note: for local types consider adding `#[derive(serde::Deserialize)]` to your `MyStruct` type
   = note: for types from other crates check whether the crate offers a `serde` feature flag
   = help: the following other types implement trait `Deserialize<'de>`:
             &'a Path
             &'a [u8]
             &'a str
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
           and $N others
note: required by a bound in `from_str`
  --> tests/ui/unimplemented/required_locally.rs:13:8
   |
11 | fn from_str<'de, T>(_: &'de str) -> T
   |    -------- required by a bound in this function
12 | where
13 |     T: Deserialize<'de>,
   |        ^^^^^^^^^^^^^^^^ required by this bound in `from_str`
